store/sqlstore/upgrade.go

```diff
@@ -1211,3 +1212,178 @@ func upgradeDatabaseToVersion537(sqlStore *SqlStore) {
+		`
+	}
+
+	if _, err := sqlStore.GetMaster().ExecNoTimeout(updateReplyCountsQuery); err != nil {
+		mlog.Error("Error updating reply counts and lastreplyat of threads", mlog.Err(err))
+	}
+	if _, err := sqlStore.GetMaster().ExecNoTimeout(updateParticipantsQuery); err != nil {
+		mlog.Error("Error updating participants of threads", mlog.Err(err))
+	}
+	if _, err := sqlStore.GetMaster().ExecNoTimeout(updateThreadMembershipQuery); err != nil {
+		mlog.Error("Error updating lastviewedat and unreadmentions of threadmemberships", mlog.Err(err))
+	}
+}
+
+func fixCRTChannelMembershipCounts(sqlStore *SqlStore) {
+	now := fmt.Sprintf("%d", model.GetMillis())
+
+	channelMembershipsCountsAndMentions := `
+		UPDATE ChannelMembers
+		SET MentionCount=0, MentionCountRoot=0, MsgCount=Channels.TotalMsgCount, MsgCountRoot=Channels.TotalMsgCountRoot, LastUpdateAt=` + now + `
+		FROM Channels
+		WHERE ChannelMembers.Channelid = Channels.Id AND ChannelMembers.LastViewedAt >= Channels.LastPostAt;
+	`
+
+	if sqlStore.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		channelMembershipsCountsAndMentions = `
+			UPDATE ChannelMembers
+			INNER JOIN Channels on Channels.Id = ChannelMembers.ChannelId
+			SET MentionCount=0, MentionCountRoot=0, MsgCount=Channels.TotalMsgCount, MsgCountRoot=Channels.TotalMsgCountRoot, LastUpdateAt=` + now + `
+			WHERE ChannelMembers.LastViewedAt >= Channels.LastPostAt;
+		`
+	}
+
+	if _, err := sqlStore.GetMaster().ExecNoTimeout(channelMembershipsCountsAndMentions); err != nil {
+		mlog.Error("Error updating counts and unreads for channelmemberships", mlog.Err(err))
+	}
+}
+
+func fixCRTChannelUnreadsForJoinLeaveMessages(sqlStore *SqlStore) {
+	now := fmt.Sprintf("%d", model.GetMillis())
+
+	if sqlStore.DriverName() == model.DATABASE_DRIVER_POSTGRES {
+		systemMessagesFixQuery := `
+			WITH
+			cte1 AS (SELECT ChannelId, UserId, LastViewedAt FROM ChannelMembers INNER JOIN Channels ON ChannelMembers.ChannelId = Channels.Id WHERE ChannelMembers.MsgCountRoot != Channels.TotalMsgCountRoot),
+			cte2 AS (SELECT ARRAY_AGG(DISTINCT Posts.Type) AS Typ, cte1.UserId, cte1.ChannelId FROM cte1 INNER JOIN Posts ON Posts.Channelid = cte1.ChannelId WHERE Posts.Createat > cte1.LastViewedAt GROUP BY cte1.UserId, cte1.ChannelId),
+			cte3 AS (SELECT cte2.*, TotalMsgCountRoot, TotalMsgCount FROM cte2 INNER JOIN Channels on cte2.ChannelId = Channels.Id)
+			UPDATE ChannelMembers SET MsgCountRoot = cte3.TotalMsgCountRoot, MsgCount = cte3.TotalMsgCount, LastUpdateAt = ` + now + `
+			FROM cte3 WHERE cte3.UserId = ChannelMembers.UserId AND ChannelMembers.ChannelId = cte3.ChannelId and '' != ALL (cte3.Typ)
+		`
+		if _, err := sqlStore.GetMaster().ExecNoTimeout(systemMessagesFixQuery); err != nil {
+			mlog.Error("Error updating unreads caused by join leave messages", mlog.Err(err))
+		}
+		return
+	}
+
+	if sqlStore.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		trans, err := sqlStore.GetMaster().Begin()
```

Comment:
Temporary tables only stick around for a session (connection). We use a transaction to ensure everything happens in the same session.
We could probably use joins instead of temporary tables, but I think this is much neater and easier to understand.